### DISCUSSION TITLE
Changing the number of substeps for `SplitExplicitFreeSurface`

### DISCUSSION THREAD

**Discussion by glwagner** [CliMA/Oceananigans.jl#3072] [Experimental features]
Currently the number of substeps is specified for the `SplitExplicitFreeSurface` on construction, eg

```julia
time_step = 5minutes
cfl = 0.7
gravitational_acceleration = 9.81
wave_speed = sqrt(gravitational_acceleration * grid.Lz)
c = Center()
min_Δx = minimum_xspacing(grid, c, c, c)
min_Δy = minimum_yspacing(grid, c, c, c)
Δ = 1 / sqrt(1 / min_Δx^2 + 1 / min_Δy^2)
substeps = max(Int(ceil(2 * time_step / (cfl / wave_speed * Δ))), 10)
free_surface = SplitExplicitFreeSurface(; gravitational_acceleration, substeps)
```

This ties the construction of the free surface to the specified time-step.

It's often useful to change the time-step while a simulation is running. How do we do this right now @simone-silvestri @sandreza ?

This is related to #2982, which would eliminate complications regarding changing the time-step with split explicit.

**Comment by simone-silvestri** [comment]
At the moment if you want to change the time step in a simulation you have to assess the substeps using the largest time step you want to achieve. 
In that way you can ensure the cfl condition is always satisfied. It is not a huge hit as the the barotropic step is very much negligible and usually we have an "idea" of the maximum time step size (if we have to change the time step by more than one order of magnitude in one simulation means that there is something wrong). This works because the wave speed is only constrained by the domain

We can think about a way to change time step. That requires being able to change halos on the fly for distributed which might make the code a bit more complicated/involved, but it can be possible to do

**Reply by glwagner** [reply]
A simple implementation of this feature would merely remake / reallocate the free surface solver when the time-step is changed, as is done within the model constructor.

The reason this is hard for a user to achieve is because the free surface solver is "materialized" within the model constructor; ie we do not have a user interface for building ready-to-go solvers.

**Comment by simone-silvestri** [comment]
I want to clarify that the split explicit does not set a barotropic "time step" but a _number_ of "substeps" and a fractional time step per substep (`Δτ`). The barotropic time step will be the fractional step * the baroclinic step (`Δτ ⋅ Δt`) , so the barotropic time step always adapts to the baroclinic step. This means that it is always safe to reduce the baroclinic time step but not to increase it tough.

**Reply by glwagner** [reply]
Okay, that's important. So one strategy, which hopefully only has a minor impact on performance, would be to initialize with an excessive number of substeps (anticipating that the baroclinic time may increase over the course of the simulation).

**Comment by glwagner** [comment]
It's worth mentioning that being able to change the time step is mostly important for idealized and/or regional applications, where forcing / turbulence can vary appreciably over the course of a simulation. For global applications the time-step requirements are much more constant (since maximum velocities etc are relatively constant in time when sampled over the whole globe incorporating both hemispheres, etc), with the possible exception of spin up periods.
